core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1689,6 +1689,13 @@ public void updateGlobalCheckpointOnReplica(final long globalCheckpoint) {
              * while the global checkpoint update may have emanated from the primary when we were in that state, we could subsequently move
              * to recovery finalization, or even finished recovery before the update arrives here.
              */
+            IndexShardState shardState = state();
+            if (shardState == IndexShardState.POST_RECOVERY ||
+                shardState == IndexShardState.STARTED ||
+                shardState == IndexShardState.RELOCATED) {
+                throw new AssertionError("supposedly in-sync shard copy received a global checkpoint [" + globalCheckpoint + "] " +
```

Comment:
This is entirely too harsh, this will fail the node if we get this wrong. We should fail the shard for sure though.